L1Trigger/Phase2L1ParticleFlow/interface/jetmet/L1PFMetEmulator.h

```diff
@@ -0,0 +1,149 @@
+#include "nlohmann/json.hpp"
+
+#ifndef CMSSW_GIT_HASH
+#include "hls_math.h"
+#endif
+
+#include <vector>
+#include <array>
+#include <fstream>
+#include <numeric>
+#include <algorithm>
+#include "ap_int.h"
+#include "ap_fixed.h"
+
+namespace L1METEmu {
+  // Define Data types for P2 L1 MET Emulator
+  typedef l1ct::pt_t pt_t;
+  typedef l1ct::glbphi_t phi_t;
+  typedef l1ct::glbeta_t eta_t;
+  typedef l1ct::dpt_t pxy_t;
+  typedef l1ct::Sum Met;
+
+  typedef ap_fixed<22, 12> proj_t;
+  typedef ap_fixed<32, 22> proj2_t;
+  typedef ap_fixed<32, 2> poly_t;
+
+  struct Particle_xy {
+    // 44bits
+    proj_t hwPx;
+    proj_t hwPy;
+  };
+
+  struct Poly2_param {
+    std::array<poly_t, 16> c0, c1, c2, s0, s1, s2;
+    std::array<phi_t, 17> phi_edges;
+  };
+
+  inline const Poly2_param& Get_Poly2_param() {
+    static Poly2_param P = [] {
+      Poly2_param t{};
+      edm::FileInPath f("L1Trigger/Phase2L1ParticleFlow/data/met/l1met_ptphi2pxpy_poly2_v1.json");
+      std::ifstream in(f.fullPath());
+      if (!in)
+        throw cms::Exception("FileNotFound") << f.fullPath();
+      nlohmann::json j;
+      in >> j;
+
+      for (int i = 0; i < 16; ++i) {
+        t.c0[i] = poly_t(j["cos"]["par0"][i].get<double>());
+        t.c1[i] = poly_t(j["cos"]["par1"][i].get<double>());
+        t.c2[i] = poly_t(j["cos"]["par2"][i].get<double>());
+        t.s0[i] = poly_t(j["sin"]["par0"][i].get<double>());
+        t.s1[i] = poly_t(j["sin"]["par1"][i].get<double>());
+        t.s2[i] = poly_t(j["sin"]["par2"][i].get<double>());
+      }
+      for (int i = 0; i < 17; ++i) {
+        t.phi_edges[i] = l1ct::Scales::makeGlbPhi(j["phi_edges"][i].get<double>() * M_PI);
+      }
+      return t;
+    }();
+    return P;
+  }
+
+  inline Particle_xy Get_xy(const l1ct::pt_t hwPt, const l1ct::glbphi_t hwPhi) {
+    /*
+      Convert pt, phi to px, py
+      - Use 2nd order Polynomial interpolation for cos, sin
+      - Divide the sine and cosine value from -2 pi to 2 pi into 16 parts
+      - Fitting the value with 2nd order function
+    */
+
+    const auto& P = L1METEmu::Get_Poly2_param();
+    int phibin = 0;
+    for (int i = 0; i < 16; i++) {
+      if (hwPhi >= P.phi_edges[i] && hwPhi < P.phi_edges[i + 1]) {
+        phibin = i;
+        break;
+      }
+    }
+
+    Particle_xy proj_xy;
+
+    poly_t cos_var = P.c0[phibin] + P.c1[phibin] * (hwPhi - P.phi_edges[phibin]) +
+                     P.c2[phibin] * (hwPhi - P.phi_edges[phibin]) * (hwPhi - P.phi_edges[phibin]);
+    poly_t sin_var = P.s0[phibin] + P.s1[phibin] * (hwPhi - P.phi_edges[phibin]) +
+                     P.s2[phibin] * (hwPhi - P.phi_edges[phibin]) * (hwPhi - P.phi_edges[phibin]);
+
+    proj_xy.hwPx = hwPt * cos_var;
+    proj_xy.hwPy = hwPt * sin_var;
+
+    return proj_xy;
+  }
+
+  inline void Sum_Particles(const std::vector<Particle_xy> particles_xy, Particle_xy& met_xy) {
+    met_xy.hwPx = 0;
+    met_xy.hwPy = 0;
+
+    for (uint i = 0; i < particles_xy.size(); ++i) {
```

Comment:
Please use `unsigned int` here and https://github.com/cms-sw/cmssw/blob/88d6bf5ea35b98871e09d3f72abf0ac0acba3a41/L1Trigger/Phase2L1ParticleFlow/interface/jetmet/L1PFMetEmulator.h#L139